### PR TITLE
remember last block so exceptions can continue where they left after the exception condition is over

### PR DESCRIPTION
**Pull request by dewagter**

### CODE REVIEW

File: sw/airborne/subsystems/navigation/common_flight_plan.h [hunk 1]
```diff
@@ -55,6 +55,7 @@ void nav_goto_block(uint8_t block_id);
 #define Label(x) label_ ## x:
 #define Goto(x) { goto label_ ## x; }
 #define Return() { nav_block=last_block; nav_stage=last_stage; block_time=0;}
+#define ReturnToStartPrevBlock() { nav_block=last_block; nav_stage=0; block_time=0;}
```
Comment (gautierhattenberger) [inline sw/airborne/subsystems/navigation/common_flight_plan.h]: The `Return` instruction can be used from a special flight plan instruction. Shall we do the same for this, or with an attribute of the existing instruction ?

File: sw/airborne/subsystems/navigation/common_flight_plan.h [hunk 2]
```diff
@@ -44,8 +44,8 @@ void nav_goto_block(uint8_t block_id);
 #define InitStage() nav_init_stage();
 
 #define Block(x) case x: nav_block=x;
-#define NextBlock() { nav_block++; nav_init_block(); }
-#define GotoBlock(b) { nav_block=b; nav_init_block(); }
+#define NextBlock() nav_goto_block(nav_block + 1)
```
Comment (podhrmic) [inline sw/airborne/subsystems/navigation/common_flight_plan.h]: Are we checking here for overflow in the number of blocks? E.g. I am in the last block and will want to go to the next block?

Comment (gautierhattenberger) [inline sw/airborne/subsystems/navigation/common_flight_plan.h]: The flight plan will fall into the default case which currently does nothing. But the instruction is not supposed to be called by hand, only from the generated code.

File: sw/airborne/subsystems/navigation/common_flight_plan.h (hunk 1 above)
Comment (dewagter) [inline sw/airborne/subsystems/navigation/common_flight_plan.h]: Would be great. Any idea which should be the default: return to last_stage or stage=0

**Comment by dewagter** [comment]
Unified: ```<return/>``` and ```<return reset_stage="1"/>``` are proposed instead of ```<call fun="ReturnToStartOfBlock"/>```

File: sw/tools/generators/gen_flight_plan.ml [hunk 3]
```diff
@@ -324,7 +324,7 @@ let rec print_stage = fun index_of_waypoints x ->
     match String.lowercase (Xml.tag x) with
         "return" ->
           stage ();
-          lprintf "Return();\n";
+          lprintf "Return(%s);\n" (ExtXml.attrib x "reset_stage");
```
Comment (gautierhattenberger) [inline sw/tools/generators/gen_flight_plan.ml]: You should not rely on the DTD to be sure that `reset_stage` attribute will exist. Better use `ExtXml.attrib_or_default` function instead.

File: sw/airborne/subsystems/navigation/common_flight_plan.h [hunk 4]
```diff
@@ -54,7 +54,7 @@ void nav_goto_block(uint8_t block_id);
 
 #define Label(x) label_ ## x:
 #define Goto(x) { goto label_ ## x; }
-#define Return() { nav_block=last_block; nav_stage=last_stage; block_time=0;}
+#define Return(RESET_STAGE) { nav_block=last_block; if (RESET_STAGE==1) {nav_stage=0;} else {nav_stage=last_stage;} block_time=0;}
```
Comment (gautierhattenberger) [inline sw/airborne/subsystems/navigation/common_flight_plan.h]: why uppercase here ?

**Pull request merged**